### PR TITLE
PHP Review

### PR DESCRIPTION
**Pull request by tfrommen**
## Description
After a quick discussion with @liamdefty where he said that WP Cypress, by default, is running on the Docker container on PHP 7.3, I made several improvements to the PHP codebase.

## Change Log
* Add parameter type declarations.
* Add return type declarations.
* Align and fix namespace usage (remove leading slashes, remove imports in the current namespace etc.).
* Use `__DIR__` instead of `dirname( __FILE__ )`.
* Use short array syntax.
* Fix and align file loading (use `require` in templates, use `include(_once)` if guarded behind an `if` etc.).
* Remove `rest_authentication_errors` filter as it was always returning the first argument no matter what?! Was that a leftover, or what was the idea?
* Some formatting (e.g., whitespace).
* Add minimum PHPDoc for class properties.
 * Use `is_readable` instead of `file_exists`.
* Import the `make_progress_bar` function from WP-CLI.
* Prevent two potential PHP notices.
* Allow for constructor injection of some dependencies. Also useful for (integration) testing.
* Split running the `run` method out of the `Seeder` constructor.
* Move the `try`-`catch` block from the `Seeder` to the `Command` class, which ideally would be the only entity to know about WP-CLI.

## Notes
All of the above, while a rather long list, is only minimal changes to the existing codebase.

That said, there are other things one could consider. I did not include any of that here, because that would make the PR much more complicated and also change the nature of it.

Some random questions that I have:
* Why is `Validation` a (static) class? The three existing methods don't share any code or dependencies, and there are no (static) class properties either. Having three namespaced methods (e.g., `WP_Cypress\validate_plugin`) would be just as good...?
* Instead of the manual `require_once` calls in `plugin.php` and `bootstrap.php`, you could add the files to the `composer.json` file, and make use of autoloading.
* What is the idea behind the `Date` trait? I mean, it is being used for both the `Seed` and the `Seeder` abstract class, but no interface defines the `now` method. One could add a `DateInterface` to define the `now` method, and then have both the `SeedInterface` and the `SeederInterface` extend that, or provide `now` as a (namespaced) function.

How would you imagine someone to enhance or add to the functionality provided by WP Cypress? For instance, if I wanted to provide a generator for terms?
At the moment, the seeder instance is created by WP Cypress, not allowing for any constructor arguments. This is OK, because you instantiate dependencies in the (abstract) constructor. But that's only true for the existing generators/seeds.
One way to allow for that, which would mean an extensive rewrite, is to make use of a container where generator implementations can be registered. In the simplest form, that "container" could be an `apply_filter` call, passing an array with seed names as keys and class instances (or maybe even /alsoclosures returning a class instance) as values.

## Types of changes (_if applicable_):

- [ ] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

## Checklist (_if applicable_):
- [ ] Meets provided linting standards.

### CODE REVIEW

**Comment by liamdefty** [comment]
Thanks for this @tfrommen. A lot of really valuable insight here and a lot of these issues were overlooked in order to get **something** working asap.

In regards to all the changes, I have no issues with any of them - PHP isn't my strongest language so I definitely appreciate the input!

There is only one thing in which I need to investigate further:

> Remove `rest_authentication_errors` filter as it was always returning the first argument no matter what?! Was that a leftover, or what was the idea?

I do remember spotting issues previously, in which this was added as a workaround and therefore was leftover and meant to be revisited. However, I am currently looking to change how the auth works which address the open issues: #25, #17 meaning this code will likely become irrelevant anyway.

---

To address some of your questions:

> Why is `Validation` a (static) class? The three existing methods don't share any code or dependencies, and there are no (static) class properties either. Having three namespaced methods (e.g., `WP_Cypress\validate_plugin`) would be just as good...?

I think initially when building this class I assumed there might of been some shared code, however that didn't end up to be the case. I will refactor to namespaced methods. (I think it makes sense to do on this branch, so I'll just add to what you've done)

-----

> Instead of the manual `require_once` calls in `plugin.php` and `bootstrap.php`, you could add the files to the `composer.json` file, and make use of autoloading.

This sounds good! If this is a relatively small change would you be able to add this to this pull request I'm not too familiar with how that would work. I'd be interested to see it though!

-----

> What is the idea behind the `Date` trait? I mean, it is being used for both the `Seed` and the `Seeder` abstract class, but no interface defines the `now` method. One could add a `DateInterface` to define the `now` method, and then have both the `SeedInterface` and the `SeederInterface` extend that, or provide `now` as a (namespaced) function.

I think initially I expected this trait to contain more than a single method. Since that isn't the case we can explore some of the options you suggested. However I think possibly having a namespace for utility functions would be the simplest.

---

> How would you imagine someone to enhance or add to the functionality provided by WP Cypress?

This is something that has been on my mind too and have envisioned something similar. I think in order for this package to be really useful this is something that needs adding. I'm going to explore a couple of options this week. 

I'd also thought of being able to define your own generator class, which extends from ours. But I don't want to make things too complicated. 

If you have any more thoughts, or would like to explore something yourself - please feel free :)

**Comment by tfrommen** [comment]
> > Remove `rest_authentication_errors` filter as it was always returning the first argument no matter what?! Was that a leftover, or what was the idea?
> 
> I do remember spotting issues previously, in which this was added as a workaround and therefore was leftover and meant to be revisited. However, I am currently looking to change how the auth works which address the open issues: #25, #17 meaning this code will likely become irrelevant anyway.

Here is the very [first commit of the `rest_authentication_errors` filter](https://github.com/bigbite/wp-cypress/commit/b2a190a1437bc7526f774bce67bcd7a46da4dfcb#diff-f6d0931a789ec27a8be743236c531dc5R3-R13) function, and it hasn't been changed since (except for indentation).


> > Instead of the manual `require_once` calls in `plugin.php` and `bootstrap.php`, you could add the files to the `composer.json` file, and make use of autoloading.
> 
> This sounds good! If this is a relatively small change would you be able to add this to this pull request I'm not too familiar with how that would work. I'd be interested to see it though!

I can have a go, yes. Just give me a ping once you are done with the refactoring you wanted to do as part of this branch/PR.

**Comment by liamdefty** [comment]
@tfrommen I've added the changes we discussed and fixed a couple of minor issues - let me know your thoughts. 

Circling back to 

> How would you imagine someone to enhance or add to the functionality provided by WP Cypress?

I think this is best left for another PR once this has been merged.

Let me know once you've had a look at autoloading - cheers!

File: plugin/Seeder/Command.php [hunk 1]
```diff
@@ -2,8 +2,8 @@
 
 namespace WP_Cypress\Seeder;
 
-use Exception;
 use WP_CLI;
+use Exception;
```
Comment (tfrommen) [inline plugin/Seeder/Command.php:6]: @liamdefty just out of curiosity, did your IDE order these? 🤔

File: plugin/inc/Validation.php [hunk 2]
```diff
@@ -1,79 +1,96 @@
 <?php
```
Comment (tfrommen) [inline plugin/inc/Validation.php:1]: Just noting that this file is still capitalized, `Validation.php`.

File: plugin/Seeder/Command.php (hunk 1 above)
Comment (liamdefty) [reply plugin/Seeder/Command.php:5]: Oops, not sure what happened there. Have re-ordered alphabetically.

**Pull request merged**